core/modules/xrdsvc/ChannelStream.cc

```diff
@@ -119,12 +93,13 @@ ChannelStream::GetBuff(XrdSsiErrInfo &eInfo, int &dlen, bool &last) {
         eInfo.Set("Not an active stream", EOPNOTSUPP);
         return 0;
     }
-    SimpleBuffer* sb = new SimpleBuffer(_msgs.front());
-    dlen = _msgs.front().size();
+
+    StreamBuffer::Ptr sb = _msgs.front();
+    dlen = sb->getSize();
     _msgs.pop_front();
     last = _closed && _msgs.empty();
     LOGS(_log, LOG_LVL_DEBUG, "returning buffer (" << dlen << ", " << (last ? "(last)" : "(more)") << ")");
-    return sb;
+    return sb.get();
```

Comment:
What about ownership here, pointer is returned but shared_ptr is deleted, is it guaranteed not to be last shared_ptr?

Reply:
The XrdSsi call takes a raw pointer to a XrdSsiStream::Buffer and calls XrdSsiStream::Buffer::Recycle when the buffer can be deleted. StreamBuffer sb has a self referential shared pointer that gets reset when Recycle is called. In some cases, other classes also have shared_ptr's to sb and can wait for Recycle to be called.